language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
             unfoldFixOf (Normalized vPat) (Normalized vArg) k
         _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
 
+
+-- Typecheck a builtin application.  This performs a lot of
+-- substitutions and normalisations, but it should preserve the global
+-- uniqueness condition because it only uses uniquness-preserving
+-- operations from Normalize.Internal.
+
+-- [infer| bn : [a_1 :: K_1, ..., a_n :: K_m](B_1, ..., B_n)C]
+-- [check| G !- A_i :: K_i]
```

Comment:
This is not what is happening right now, because currently we assume that all `K_i` are `*`, right?

Reply:
Yes, but the specification allows higher kinds. I'm not sure what to do there.  This relates to the remark above. Maybe we should make the builtin signatures carry kinds as well so that we can ignore that issue here and localise the problem in one place.  At the moment we have a global assumption that everything has kind `*` and it would be easy to change that in one place and forget to fix it elsewhere.